File: sdk/golang/tests/e2e/ws/spot_test/public_test.go
package spot_test

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"sync"
	"testing"

	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/api"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/common/logger"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/generate/spot/spotpublic"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/types"
	"github.com/stretchr/testify/assert"
)

var spotPublicApi spotpublic.SpotPublicWS

func init() {
	logger.SetLogger(
		&logger.DefaultLogger{
			Logger:     log.New(os.Stdout, "", 0),
			Level:      logger.TRACE,
			Formatter:  "%s [%s] [%s+%d] [%s] %s",
			TimeFormat: "2006-01-02 15:04:05.000",
		})

	// get api key from env
	key := os.Getenv("API_KEY")
	secret := os.Getenv("API_SECRET")
	passphrase := os.Getenv("API_PASSPHRASE")

	// set callback worker number
	wsOption := types.NewWebSocketClientOptionBuilder().
		Build()

	// create a client with transport options
	option := types.NewClientOptionBuilder().
		WithKey(key).
		WithSecret(secret).
		WithPassphrase(passphrase).
		WithSpotEndpoint(types.GlobalApiEndpoint).
		WithFuturesEndpoint(types.GlobalFuturesApiEndpoint).
		WithBrokerEndpoint(types.GlobalBrokerApiEndpoint).
		WithWebSocketClientOption(wsOption).
		Build()

	client := api.NewClient(option)

	spotPublicApi = client.WsService().NewSpotPublicWS()
	err := spotPublicApi.Start()
	if err != nil {
		panic(err)
	}
}

func TestPublicAllTickers(t *testing.T) {
	wg := sync.WaitGroup{}
	wg.Add(1)

	i := 0
	id, err := spotPublicApi.AllTickers(func(topic string, subject string, data *spotpublic.AllTickersEvent) error {
		assert.NotNil(t, data)
		i++
		if i == 1 {
			wg.Done()
		}
		str, _ := json.Marshal(data)
		fmt.Println(topic, subject, string(str))
		return nil
	})
	assert.Nil(t, err)
	wg.Wait()

	err = spotPublicApi.UnSubscribe(id)
	assert.Nil(t, err)
}

func TestPublicKlines(t *testing.T) {
	wg := sync.WaitGroup{}
	wg.Add(1)

	i := 0
	id, err := spotPublicApi.Klines("BTC-USDT", "1min", func(topic string, subject string, data *spotpublic.KlinesEvent) error {
		assert.NotNil(t, data)
		i++
		if i == 10 {
			wg.Done()
		}
		str, _ := json.Marshal(data)
		fmt.Println(topic, subject, string(str))
		return nil
	})
	assert.Nil(t, err)
	wg.Wait()

	err = spotPublicApi.UnSubscribe(id)
	assert.Nil(t, err)
}

func TestPublicMarketSnapshot(t *testing.T) {
	wg := sync.WaitGroup{}
	wg.Add(1)

	i := 0
	id, err := spotPublicApi.MarketSnapshot("BTC-USDT", func(topic string, subject string, data *spotpublic.MarketSnapshotEvent) error {
		assert.NotNil(t, data)
		i++
		if i == 10 {
			wg.Done()
		}
		str, _ := json.Marshal(data)
		fmt.Println(topic, subject, string(str))
		return nil
	})
	assert.Nil(t, err)
	wg.Wait()

	err = spotPublicApi.UnSubscribe(id)
	assert.Nil(t, err)
}

func TestPublicOrderbookIncrement(t *testing.T) {
	wg := sync.WaitGroup{}
	wg.Add(1)

	i := 0
	id, err := spotPublicApi.OrderbookIncrement([]string{"BTC-USDT", "DOGE-USDT"}, func(topic string, subject string, data *spotpublic.OrderbookIncrementEvent) error {
		assert.NotNil(t, data)
		i++
		if i == 10 {
			wg.Done()
		}
		str, _ := json.Marshal(data)
		fmt.Println(topic, subject, string(str))
		return nil
	})
	assert.Nil(t, err)
	wg.Wait()

	err = spotPublicApi.UnSubscribe(id)
	assert.Nil(t, err)
}

func TestPublicOrderbookLevel1(t *testing.T) {
	wg := sync.WaitGroup{}
	wg.Add(1)

	i := 0
	id, err := spotPublicApi.OrderbookLevel1([]string{"BTC-USDT", "DOGE-USDT"}, func(topic string, subject string, data *spotpublic.OrderbookLevel1Event) error {
		assert.NotNil(t, data)
		i++
		if i == 10 {
			wg.Done()
		}
		str, _ := json.Marshal(data)
		fmt.Println(topic, subject, string(str))
		return nil
	})
	assert.Nil(t, err)
	wg.Wait()

	err = spotPublicApi.UnSubscribe(id)
	assert.Nil(t, err)
}

func TestPublicOrderbookLevel50(t *testing.T) {
	wg := sync.WaitGroup{}
	wg.Add(1)

	i := 0
	id, err := spotPublicApi.OrderbookLevel50([]string{"BTC-USDT", "DOGE-USDT"}, func(topic string, subject string, data *spotpublic.OrderbookLevel50Event) error {
		assert.NotNil(t, data)
		i++
		if i == 10 {
			wg.Done()
		}
		str, _ := json.Marshal(data)
		fmt.Println(topic, subject, string(str))
		return nil
	})
	assert.Nil(t, err)
	wg.Wait()

	err = spotPublicApi.UnSubscribe(id)
	assert.Nil(t, err)
}

func TestPublicOrderbookLevel5(t *testing.T) {
	wg := sync.WaitGroup{}
	wg.Add(1)

	i := 0
	id, err := spotPublicApi.OrderbookLevel5([]string{"BTC-USDT", "DOGE-USDT"}, func(topic string, subject string, data *spotpublic.OrderbookLevel5Event) error {
		assert.NotNil(t, data)
		i++
		if i == 10 {
			wg.Done()
		}
		str, _ := json.Marshal(data)
		fmt.Println(topic, subject, string(str))
		return nil
	})
	assert.Nil(t, err)
	wg.Wait()

	err = spotPublicApi.UnSubscribe(id)
	assert.Nil(t, err)
}

func TestPublicSymbolSnapshot(t *testing.T) {
	wg := sync.WaitGroup{}
	wg.Add(1)

	i := 0
	id, err := spotPublicApi.SymbolSnapshot("BTC-USDT", func(topic string, subject string, data *spotpublic.SymbolSnapshotEvent) error {
		assert.NotNil(t, data)
		i++
		if i == 10 {
			wg.Done()
		}
		str, _ := json.Marshal(data)
		fmt.Println(topic, subject, string(str))
		return nil
	})
	assert.Nil(t, err)
	wg.Wait()

	err = spotPublicApi.UnSubscribe(id)
	assert.Nil(t, err)
}

func TestPublicTicker(t *testing.T) {
	wg := sync.WaitGroup{}
	wg.Add(1)

	i := 0
	id, err := spotPublicApi.Ticker([]string{"BTC-USDT", "ETH-USDT"}, func(topic string, subject string, data *spotpublic.TickerEvent) error {
		assert.NotNil(t, data)
		i++
		if i == 10 {
			wg.Done()
		}
		str, _ := json.Marshal(data)
		fmt.Println(topic, subject, string(str))
		return nil
	})
	assert.Nil(t, err)
	wg.Wait()

	err = spotPublicApi.UnSubscribe(id)
	assert.Nil(t, err)
}

func TestPublicTrade(t *testing.T) {
	wg := sync.WaitGroup{}
	wg.Add(1)

	i := 0
	id, err := spotPublicApi.Trade([]string{"BTC-USDT", "ETH-USDT"}, func(topic string, subject string, data *spotpublic.TradeEvent) error {
		assert.NotNil(t, data)
		i++
		if i == 10 {
			wg.Done()
		}
		str, _ := json.Marshal(data)
		fmt.Println(topic, subject, string(str))
		return nil
	})
	assert.Nil(t, err)
	wg.Wait()

	err = spotPublicApi.UnSubscribe(id)
	assert.Nil(t, err)
}

File: sdk/python/test/e2e/ws/spot_test/public_test.py
import logging
import os
import threading
import unittest

from kucoin_universal_sdk.api.client import DefaultClient
from kucoin_universal_sdk.model.client_option import ClientOptionBuilder
from kucoin_universal_sdk.model.constants import GLOBAL_API_ENDPOINT, GLOBAL_FUTURES_API_ENDPOINT, \
    GLOBAL_BROKER_API_ENDPOINT
from kucoin_universal_sdk.model.websocket_option import WebSocketClientOptionBuilder

logging.basicConfig(level=logging.DEBUG)


class SpotPublicWsTest(unittest.TestCase):
    def setUp(self):
        key = os.getenv("API_KEY")
        secret = os.getenv("API_SECRET")
        passphrase = os.getenv("API_PASSPHRASE")

        if not key or not secret or not passphrase:
            self.fail("Environment variables API_KEY, API_SECRET, and API_PASSPHRASE must be set.")

        # Create client option
        ws_client_option = WebSocketClientOptionBuilder().build()
        client_option = (
            ClientOptionBuilder()
            .set_key(key)
            .set_secret(secret)
            .set_passphrase(passphrase)
            .set_websocket_client_option(ws_client_option)
            .set_spot_endpoint(GLOBAL_API_ENDPOINT)
            .set_futures_endpoint(GLOBAL_FUTURES_API_ENDPOINT)
            .set_broker_endpoint(GLOBAL_BROKER_API_ENDPOINT)
            .build()
        )

        # Create WebSocket client
        client = DefaultClient(client_option)
        kucoin_ws_client = client.ws_service()
        self.api = kucoin_ws_client.new_spot_public_ws()
        self.api.start()

    def tearDown(self):
        self.api.stop()

    def _test_subscription(self, subscribe_method, *args):
        message_count = 0
        event = threading.Event()

        def callback(topic: str, subject: str, data):
            nonlocal message_count
            message_count += 1
            print(f"Received message {message_count} on topic: {topic}, subject: {subject}, data: {data.to_json()}")
            if message_count == 10:
                print("Received 10 messages. Unsubscribe...")
                event.set()

        subscription_id = subscribe_method(*args, callback)
        try:
            event.wait()
        finally:
            self.api.unsubscribe(subscription_id)

    def test_public_all_tickers(self):
        self._test_subscription(self.api.all_tickers)

    def test_public_tickers(self):
        symbols = ["BTC-USDT", "ETH-USDT"]
        self._test_subscription(self.api.ticker, symbols)

    def test_public_klines(self):
        self._test_subscription(self.api.klines, "BTC-USDT", "1min")

    def test_public_trade(self):
        symbols = ["BTC-USDT", "ETH-USDT"]
        self._test_subscription(self.api.trade, symbols)

    def test_public_market_snapshot(self):
        symbols = "BTC-USDT"
        self._test_subscription(self.api.market_snapshot, symbols)

    def test_public_symbol_snapshot(self):
        symbols = "BTC-USDT"
        self._test_subscription(self.api.symbol_snapshot, symbols)

    def test_public_orderbook_lever1(self):
        symbols = ["BTC-USDT", "ETH-USDT"]
        self._test_subscription(self.api.orderbook_level1, symbols)

    def test_public_orderbook_lever5(self):
        symbols = ["BTC-USDT", "ETH-USDT"]
        self._test_subscription(self.api.orderbook_level5, symbols)

    def test_public_orderbook_lever50(self):
        symbols = ["BTC-USDT", "ETH-USDT"]
        self._test_subscription(self.api.orderbook_level50, symbols)

    def test_public_orderbook_increment(self):
        symbols = ["BTC-USDT", "ETH-USDT"]
        self._test_subscription(self.api.orderbook_increment, symbols)
